Have only one source where testsuites are defined

Marenz opened the pull request.

Marenz committed:
Have only one source where testsuits are defined

ekpyron commented on test/testsuits.h:
```diff
@@ -0,0 +1,49 @@
+
+#include <boost/filesystem.hpp>
+
+using namespace dev::solidity::test;
```
I think you should rather make everything in the header part of this namespace - we usually doesn't import namespaces in header files. You can import it in the implementation files (if that's not already done anyways).

ekpyron reviewed (changes requested):
Looks fine apart from the namespace import!

chriseth commented on test/testsuits.h:
```diff
@@ -0,0 +1,49 @@
+};
+
+
+/// Array of all testsuits
```
This is just the list of test suites that are usable with `isoltest`. I think the name of this file should also reflect that. Perhaps `InteractiveTests.h`?

ekpyron replied:
``InteractiveTests.h`` sounds more like the tests are *only* to be run interactively.

It's all extracted tests, isn't it? So if we want to be more specific than just "testsuits", then maybe that's the way to go, although ``ExtractedTests.h`` doesn't sound good either IMO, so not sure, maybe there's something else that's better...

chriseth replied:
Another thing they have in common is that they generate expectations that are compared. They are also auto-updatable. Just some inspiration :)

Marenz replied:
"extracted" implies that they are not in their "normal" place but put there, taken from their normal environment. That seems neither correct nor important in this context? 

Why are you unhappy with `testsuites.h`?

chriseth replied:
`testsuites.h` is too generic, because not all test suites can be covered by this.

ekpyron replied:
Maybe ``TestCaseDirectories``? I'm terrible at naming things ;-). But since the base class for the extracted tests is named ``TestCase`` anyways (although it's again not *all* of the test cases...)...

lazaridiscom commented:
Not sure why you don't simply apply the reality-tested #5563 (and ideally #5580), and instead go on with work which essentially keeps up the coupling.

In any way: can you please close #5563 and #5580?

Marenz commented:
@lazaridiscom as far as I can tell #5563 + #5580 and this PR don't contradict each other. Both can be applied. They simply have different goals.

lazaridiscom commented:
Your work overlaps my already started/finalized work (especially #5563), so there is no reason to keep my work up.

I have anyways rewritten my use-case, so I do not depend on those PRs getting merged.

So, just close them , and go on.

Marenz commented:
Update:
- Using the test namespace now
- Renamed to `InteractiveTests.h` after consulting with @chriseth 
- Added a few comments explaining the structure

Marenz commented:
Aaaand rebased.

chriseth commented on test/InteractiveTests.h:
```diff
@@ -0,0 +1,61 @@
+/** Container for all information regarding a testsuite */
+struct Testsuite
+{
+	const char* title;
```
```suggestion
	char const* title;
```

chriseth commented on test/InteractiveTests.h:
```diff
@@ -0,0 +1,61 @@
+struct Testsuite
+{
+	const char* title;
+	const boost::filesystem::path path;
```
```suggestion
	boost::filesystem::path const path;
```

chriseth commented on test/InteractiveTests.h:
```diff
@@ -0,0 +1,61 @@
+{
+	const char* title;
+	const boost::filesystem::path path;
+	const boost::filesystem::path subpath;
```
```suggestion
	boost::filesystem::path const subpath;
```

chriseth commented on test/boostTest.cpp:
```diff
@@ -129,46 +126,26 @@ test_suite* init_unit_test_suite( int /*argc*/, char* /*argv*/[] )
-	if (!dev::test::Options::get().disableSMT)
+
+	// Include the interactive tests in the automatic tests as well
+	for (const auto& ts : testsuits)
```
```suggestion
	for (const auto& ts: testsuits)
```

chriseth commented on test/InteractiveTests.h:
```diff
@@ -0,0 +1,61 @@
+
+/// Array of testsuits that can be run interactively as well as automatically
+const Testsuite testsuits[] = {
+	{"Yul Optimizer",       "libyul",      "yulOptimizerTests",   false, false, &yul::test::YulOptimizerTest::create},
```
Could you add a line of comment with column headers (especially for the bools)?

chriseth commented on test/boostTest.cpp:
```diff
@@ -129,46 +126,26 @@ test_suite* init_unit_test_suite( int /*argc*/, char* /*argv*/[] )
-			"smtCheckerTests",
-			SyntaxTest::create
-		) > 0, "no SMT checker tests found");
+		const auto& options = dev::test::Options::get();
```
```suggestion
		auto const& options = dev::test::Options::get();
```

chriseth commented on test/InteractiveTests.h:
```diff
@@ -0,0 +1,61 @@
+
+
+/// Array of testsuits that can be run interactively as well as automatically
+const Testsuite testsuits[] = {
```
```suggestion
Testsuite const g_testsuites[] = {
```

chriseth replied:
or even `g_interactiveTestSuites`?

Marenz replied:
or `g_interactiveTests`?

Marenz commented:
Updated:
* renamed global array to `g_interactiveTestsuites`
* Added column explaining comment
* further fixes to adhere to coding guidelines (thank you for the commit suggestions)
